File: packages/walletconnect_pay/android/src/main/kotlin/com/walletconnect/walletconnect_pay/WalletConnectPayClient.kt
package com.walletconnect.walletconnect_pay

import android.util.Log
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import uniffi.yttrium_wcpay.PayJsonException

import uniffi.yttrium_wcpay.WalletConnectPayJson

class WalletConnectPayClient {
    companion object {
        private lateinit var walletConnectPayClient: WalletConnectPayJson

        private fun handlePayJsonException(e: PayJsonException, result: MethodChannel.Result) {
            val errorCode = e::class.simpleName ?: "PayError"
            result.error(errorCode, e.message, null)
        }

        fun initialize(params: Any?, result: MethodChannel.Result) {
            val sdkConfig = params as? String ?: return result.error("PayError", "Invalid init parameters: $params", null)
            Log.d("🤖 WalletConnectPayClient.initialize", "sdkConfig: $sdkConfig")

            try {
                walletConnectPayClient = WalletConnectPayJson(sdkConfig)
                result.success(true)
            } catch (e: PayJsonException) {
                handlePayJsonException(e, result)
            } catch (e: Exception) {
                Log.d("🤖 WalletConnectPayClient.initialize", "❌: ${e.message}")
                result.error("PayError", e.message, null)
            }
        }

        fun getPaymentOptions(params: Any?, result: MethodChannel.Result) {
            val requestJson = params as? String ?: return result.error("PayError", "Invalid getPaymentOptions parameters: $params", null)
            Log.d("🤖 WalletConnectPayClient.getPaymentOptions", "requestJson: $requestJson")

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val createPaymentResponse = walletConnectPayClient.getPaymentOptions(requestJson)
                    result.success(createPaymentResponse)
                } catch (e: PayJsonException) {
                    handlePayJsonException(e, result)
                } catch (e: Exception) {
                    Log.d("🤖 WalletConnectPayClient.getPaymentOptions", "❌: ${e.message}")
                    result.error("PayError", e.message, null)
                }
            }

        }

        fun getRequiredPaymentActions(params: Any?, result: MethodChannel.Result) {
            val requestJson = params as? String ?: return result.error("PayError", "Invalid getRequiredPaymentActions parameters: $params", null)
            Log.d("🤖 WalletConnectPayClient.getRequiredPaymentActions", "requestJson: $requestJson")

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val createPaymentResponse = walletConnectPayClient.getRequiredPaymentActions(requestJson)
                    result.success(createPaymentResponse)
                } catch (e: PayJsonException) {
                    handlePayJsonException(e, result)
                } catch (e: Exception) {
                    Log.d("🤖 WalletConnectPayClient.getRequiredPaymentActions", "❌: ${e.message}")
                    result.error("PayError", e.message, null)
                }
            }

        }

        fun confirmPayment(params: Any?, result: MethodChannel.Result) {
            val requestJson = params as? String ?: return result.error("PayError", "Invalid confirmPayment parameters: $params", null)
            Log.d("🤖 WalletConnectPayClient.confirmPayment", "requestJson: $requestJson")

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val createPaymentResponse = walletConnectPayClient.confirmPayment(requestJson)
                    result.success(createPaymentResponse)
                } catch (e: PayJsonException) {
                    handlePayJsonException(e, result)
                } catch (e: Exception) {
                    Log.d("🤖 WalletConnectPayClient.confirmPayment", "❌: ${e.message}")
                    result.error("PayError", e.message, null)
                }
            }

        }
    }
}

File: packages/walletconnect_pay/ios/Classes/WalletConnectPayClient.swift
import Flutter
import Foundation
import YttriumWrapper

class WalletConnectPayClient {
    private static var walletConnectPayClient: WalletConnectPayJson?
    
    static func initialize(params: Any?, result: @escaping FlutterResult) {
        guard let sdkConfig = params as? String else {
            result(FlutterError(code: "PayError", message: "Invalid init parameters: \(String(describing: params))", details: nil))
            return
        }
        
        print("🤖 WalletConnectPayClient.initialize sdkConfig: \(sdkConfig)")
        
        do {
            walletConnectPayClient = try WalletConnectPayJson(sdkConfig: sdkConfig)
            result(true)
        } catch let e as ConfigError {
            switch e {
            case .MissingAuth:
                result(FlutterError(code: "MissingAuth", message: e.errorDescription ?? "Missing authentication", details: nil))
            }
        } catch {
            print("🤖 WalletConnectPayClient.initialize ❌: \(error.localizedDescription)")
            result(FlutterError(code: "PayError", message: error.localizedDescription, details: nil))
        }
    }
    
    static func getPaymentOptions(params: Any?, result: @escaping FlutterResult) {
        guard let requestJson = params as? String else {
            result(FlutterError(code: "PayError", message: "Invalid getPaymentOptions parameters: \(String(describing: params))", details: nil))
            return
        }
        
        print("🤖 WalletConnectPayClient.getPaymentOptions requestJson: \(requestJson)")
        
        guard let client = walletConnectPayClient else {
            result(FlutterError(code: "WalletConnectPayClient", message: "Client not initialized. Call initialize first.", details: nil))
            return
        }
        
        Task {
            do {
                let response = try await client.getPaymentOptions(requestJson: requestJson)
                result(response)
            } catch let e as GetPaymentOptionsError {
                switch e {
                case .Http(let message):
                    result(FlutterError(code: "Http", message: message, details: nil))
                case .ComplianceFailed(let message):
                    result(FlutterError(code: "ComplianceFailed", message: message, details: nil))
                case .InternalError(let message):
                    result(FlutterError(code: "InternalException", message: message, details: nil))
                case .InvalidAccount(let message):
                    result(FlutterError(code: "InvalidAccount", message: message, details: nil))
                case .InvalidRequest(let message):
                    result(FlutterError(code: "InvalidRequest", message: message, details: nil))
                case .OptionNotFound(let message):
                    result(FlutterError(code: "OptionNotFound", message: message, details: nil))
                case .PaymentExpired(let message):
                    result(FlutterError(code: "PaymentExpired", message: message, details: nil))
                case .PaymentNotFound(let message):
                    result(FlutterError(code: "PaymentNotFound", message: message, details: nil))
                case .PaymentNotReady(let message):
                    result(FlutterError(code: "PaymentNotReady", message: message, details: nil))
                }
            } catch {
                print("🤖 WalletConnectPayClient.getPaymentOptions ❌: \(error.localizedDescription)")
                result(FlutterError(code: "PayError", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    static func getRequiredPaymentActions(params: Any?, result: @escaping FlutterResult) {
        guard let requestJson = params as? String else {
            result(FlutterError(code: "PayError", message: "Invalid getRequiredPaymentActions parameters: \(String(describing: params))", details: nil))
            return
        }
        
        print("🤖 WalletConnectPayClient.getRequiredPaymentActions requestJson: \(requestJson)")
        
        guard let client = walletConnectPayClient else {
            result(FlutterError(code: "WalletConnectPayClient", message: "Client not initialized. Call initialize first.", details: nil))
            return
        }
        
        Task {
            do {
                let response = try await client.getRequiredPaymentActions(requestJson: requestJson)
                result(response)
            } catch let e as GetPaymentRequestError {
                switch e {
                case .Http(let message):
                    result(FlutterError(code: "Http", message: message, details: nil))
                case .FetchError(let message):
                    result(FlutterError(code: "FetchException", message: message, details: nil))
                case .InternalError(let message):
                    result(FlutterError(code: "InternalException", message: message, details: nil))
                case .InvalidAccount(let message):
                    result(FlutterError(code: "InvalidAccount", message: message, details: nil))
                case .OptionNotFound(let message):
                    result(FlutterError(code: "OptionNotFound", message: message, details: nil))
                case .PaymentNotFound(let message):
                    result(FlutterError(code: "PaymentNotFound", message: message, details: nil))
                }
            } catch {
                print("🤖 WalletConnectPayClient.getRequiredPaymentActions ❌: \(error.localizedDescription)")
                result(FlutterError(code: "PayError", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    static func confirmPayment(params: Any?, result: @escaping FlutterResult) {
        guard let requestJson = params as? String else {
            result(FlutterError(code: "PayError", message: "Invalid confirmPayment parameters: \(String(describing: params))", details: nil))
            return
        }
        
        print("🤖 WalletConnectPayClient.confirmPayment requestJson: \(requestJson)")
        
        guard let client = walletConnectPayClient else {
            result(FlutterError(code: "WalletConnectPayClient", message: "Client not initialized. Call initialize first.", details: nil))
            return
        }
        
        Task {
            do {
                let response = try await client.confirmPayment(requestJson: requestJson)
                result(response)
            } catch let e as ConfirmPaymentError {
                switch e {
                case .Http(let message):
                    result(FlutterError(code: "Http", message: message, details: nil))
                case .InternalError(let message):
                    result(FlutterError(code: "InternalException", message: message, details: nil))
                case .InvalidOption(let message):
                    result(FlutterError(code: "InvalidOption", message: message, details: nil))
                case .InvalidSignature(let message):
                    result(FlutterError(code: "InvalidSignature", message: message, details: nil))
                case .PaymentExpired(let message):
                    result(FlutterError(code: "PaymentExpired", message: message, details: nil))
                case .PaymentNotFound(let message):
                    result(FlutterError(code: "PaymentNotFound", message: message, details: nil))
                case .RouteExpired(let message):
                    result(FlutterError(code: "RouteExpired", message: message, details: nil))
                case .UnsupportedMethod(let message):
                    result(FlutterError(code: "UnsupportedMethod", message: message, details: nil))
                }
            } catch {
                print("🤖 WalletConnectPayClient.confirmPayment ❌: \(error.localizedDescription)")
                result(FlutterError(code: "PayError", message: error.localizedDescription, details: nil))
            }
        }
    }
}
